src/nmodl/modl.cpp

```diff
@@ -344,23 +368,73 @@ static void openfiles(char* given_filename, char* output_dir) {
 #endif
 }
 
-static std::string str_replace(std::string str, const std::string& search_str, const std::string& replace_str)
-{
-    if (search_str.empty()) {
-        return str;
-    }
-
-    size_t pos;
-    while ( (pos = str.find(search_str)) != std::string::npos ) {
-        str.replace(pos, search_str.size(), replace_str);
-    }
-
-    return str;
-}
-
 // Post-adjustments for VERBATIM blocks  (i.e  make them compatible with CPP).
 void verbatim_adjust(char* q) {
-    // template is a reserved CPP keyword
-    const std::string repl = str_replace(q, "u.template", "u.ctemplate");
+    std::string repl{q};
+    auto const regex_replace = [&repl](const char* pattern, const char* replacement) {
+        repl = std::regex_replace(std::move(repl), std::regex{pattern}, replacement);
+    };
+    auto const regex_remove = [&](const char* pattern) { regex_replace(pattern, ""); };
+    // Adjustments that aim to make old mechanisms compilable as C++. From
+    // around NEURON 8.1 this will be the default, and these transformations
+    // should not be enabled for new mechanisms.
+    if (transformations.count("cxx")) {
+        // C++ declarations must be correct; C was much sloppier and this carries
+        // over into many .mod files in the wild. Try and remove declarations
+        // from VERBATIM blocks and assume that the correct declarations of these
+        // utility functions are visible via implicitly included headers.
+
+        // These need to be fudged out because we cannot overload by return type.
+        // This is supposed to match an optionally-extern function returning void*
+        // called nrn_random_arg, vector_arg or vector_new1 and taking any number of
+        // arguments.
+        regex_remove(
+            "(extern\\s{1}|)\\s*void\\s*\\*\\s*(nrn_random_arg|vector_arg|vector_new1)\\(.*?\\);");
+        // Local declarations of double *hoc_pgetarg(void) shadow the global
+        // declaration that takes int. Transforms:
+        //   extern double hoc_call_func();
+        //   extern Object** hoc_objgetarg();
+        //   FILE* f, *hoc_obj_file_arg();
+        //   double *xdir, *xval, *hoc_pgetarg();
+        //   char *gargstr(), *filename;"
```

Comment:
Discussed this offline and understood rational for these 3 transformations but they are bit ad-hoc. Will add some notes in the review summary.

Reply:
Agree they're very ad-hoc. I think the solution is that we should not apply these transformations in any kind of open-ended on-by-default way, only to existing models in the ModelDB.